csrc/executor.cpp

```diff
@@ -1606,6 +1607,137 @@ void FusionExecutor::initializeExecutorEntry(
   executor_entry.init = true;
 }
 
+/// Copies the data, logical_size, and alloc_stride parameters to the
+/// appropriate parts of entry.args[idx].
+///
+/// For GPU tensors, we pass a Tensor<type, rank, rank> struct (see
+/// runtime/tensor.cu), where the rank describes the number of elements in the
+/// shape and stride arrays. The actual shapes and strides are dynamic, but the
+/// type and rank of the tensors are actually static (changing them would need
+/// a new FusionDefinition). So we create the storage area for the
+/// Tensor<t,r,r> during ::computeArgs, and then in this function we just
+/// update that memory with the current values for the tensor's base address,
+/// shape, and strides.
+///
+/// @param entry the entry we have previously setup for this fusion
+/// @param idx the index into entry.args and related parallel arrays in the
+///            entry.
+/// @param idx_type_size generally sizeof(int32_t) or sizeof(int64_t); used for
+///                      computing how large the arrays to copy are.
+static void fill_gpu_ptr(
+    FusionExecutor::ExecutorEntry& entry,
+    const PolymorphicValue& tmd,
+    size_t idx,
+    size_t idx_type_size) {
+  void* data = tmd->*&TensorMetaData::data;
```

Comment:
Maybe a question to @zasdfgbnm: any chance to simplify the API? It seems that something like `pv.as<TensorMetaData>().data` would be more straightforward to many C++ engineers. What's the reason behind the current API?

Reply:
`pv.as<TensorMetaData>().data` requires `TensorMetaData` to be known at the definition of `PolymorphicValue`. Basically, you will need something like:

```C++
using PolymorphicValue = dynamic_type::DynamicType<
    dynamic_type::Containers<std::vector>,
    TensorMetaData,
    ....>;
```

`TensorMetaData` is not the only struct we support, and we do not want to manually specify every struct in the definition of `PolymorphicValue`, instead we want to be able to define an arbitrary struct and let `PolymorphicValue` hold it.